### PR TITLE
Signals module

### PR DESCRIPTION
**Pull request by leunardo**

### CODE REVIEW

File: lib/jni/LogManager_jni.cpp [hunk 1]
```diff
@@ -1675,3 +1681,62 @@ Java_com_microsoft_applications_events_LogManagerProvider_00024LogManagerImpl_na
...
+        logManager->RemoveDataInspector(ss->GetName());
+        return true;
+    }
+#endif
+    return false;
+}
+
+extern "C"
+JNIEXPORT jboolean JNICALL
+Java_com_microsoft_applications_events_LogManager_nativeRegisterSubstrateSignalsOnDefaultLogManager(JNIEnv *env, jclass clazz) {
```
Comment (lalitb) [inline lib/jni/LogManager_jni.cpp:1717]: nit - we can move the `LogManager_*` functions as part of the single encapsulated `extern "C"` to be in consistent with the existing pattern.

File: lib/jni/SubstrateSignals_jni.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+JNIEXPORT jboolean JNICALL
+Java_com_microsoft_applications_events_SubstrateSignals_sendSignal(JNIEnv *env,
+                                                                              jclass clazz,
+                                                                              jlong nativeLoggerPtr,
+                                                                              jstring signal_item_json) {
+    jboolean isCopy = true;
+    const char *signalItemJson = (env)->GetStringUTFChars(signal_item_json, &isCopy);
+    env->ReleaseStringUTFChars(signal_item_json, signalItemJson);
+
+    // SubstrateSignals::CreateEventProperties(std::string json)
```
Comment (lalitb) [inline lib/jni/SubstrateSignals_jni.cpp:33]: nit - we can remove this comment ?

File: lib/android_build/maesdk/src/main/java/com/microsoft/applications/events/SubstrateSignals.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package com.microsoft.applications.events;
+
+public class SubstrateSignals {
+    /**
+     * Creates EventProperties object from json SignalItem.
```
Comment (lalitb) [inline lib/android_build/maesdk/src/main/java/com/microsoft/applications/events/SubstrateSignals.java:5]: nit - is this comment relevant to the method ?

File: lib/android_build/maesdk/src/main/java/com/microsoft/applications/events/SubstrateSignals.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+     *
+     * @param loggerPtr the ILogger pointer
+     * @param signalItemJson the Signal Item in json format.
+     */
+    public static native boolean sendSignal(long loggerPtr, String signalItemJson);
```
Comment (lalitb) [inline lib/android_build/maesdk/src/main/java/com/microsoft/applications/events/SubstrateSignals.java:10]: Just curious, how is this method going to be invoked ?

**Comment by lalitb** [comment]
Are there any plans to provide the iOS bindings for this extension ?

File: lib/android_build/maesdk/src/main/java/com/microsoft/applications/events/SubstrateSignals.java (hunk 4 above)
Comment (leunardo) [reply lib/android_build/maesdk/src/main/java/com/microsoft/applications/events/SubstrateSignals.java:10]: Hey, it will be called after an user interaction happened. The first candidate would be Microsoft Teams "LinkClicked" signal, that is generated by the client when the user clicks on a link.

**Comment by leunardo** [comment]
> Are there any plans to provide the iOS bindings for this extension ?

Not right now, we are considering as a nice-to-have addition. We might work with our partners in the future to leverage their solution on iOS and port it to the extension.

File: Solutions/MSTelemetrySDK.sln [hunk 5]
```diff
@@ -1277,8 +1197,97 @@ Global
...
+		..\lib\modules\liveeventinspector\liveeventinspector.vcxitems*{f8bb42f6-9c93-423b-b0c6-23ef79dd959a}*SharedItemsImports = 9
+		..\lib\decoder\decoder.vcxitems*{fe79fb3a-b3ef-38df-b7a4-11277db72b39}*SharedItemsImports = 4
+		..\lib\modules\azmon\azmon.vcxitems*{fe79fb3a-b3ef-38df-b7a4-11277db72b39}*SharedItemsImports = 4
+		..\lib\modules\cds\cds.vcxitems*{fe79fb3a-b3ef-38df-b7a4-11277db72b39}*SharedItemsImports = 4
+		..\lib\modules\dataviewer\dataviewer.vcxitems*{fe79fb3a-b3ef-38df-b7a4-11277db72b39}*SharedItemsImports = 4
+		..\lib\modules\filter\filter.vcxitems*{fe79fb3a-b3ef-38df-b7a4-11277db72b39}*SharedItemsImports = 4
+		..\lib\modules\liveeventinspector\liveeventinspector.vcxitems*{fe79fb3a-b3ef-38df-b7a4-11277db72b39}*SharedItemsImports = 4
+		..\lib\modules\privacyguard\privacyguard.vcxitems*{fe79fb3a-b3ef-38df-b7a4-11277db72b39}*SharedItemsImports = 4
+		..\lib\pal\desktop\desktop.vcxitems*{fe79fb3a-b3ef-38df-b7a4-11277db72b39}*SharedItemsImports = 4
+		Clienttelemetry\Clienttelemetry.vcxitems*{fe79fb3a-b3ef-38df-b7a4-11277db72b39}*SharedItemsImports = 4
```
Comment (lalitb) [inline Solutions/MSTelemetrySDK.sln:1291]: Instead of changing this whole file, can we just add the required vcx files for substrate

File: CMakeLists.txt [hunk 6]
```diff
@@ -253,6 +253,7 @@ option(BUILD_PACKAGE      "Build package"           YES)
 option(BUILD_PRIVACYGUARD "Build Privacy Guard"     YES)
 option(BUILD_CDS          "Build CDS - Common Diagnostic Stack"     YES)
 option(BUILD_LIVEEVENTINSPECTOR   "Build Live Event Inspector"      YES)
+option(BUILD_SUBSTRATESIGNALS "Build Substrate Signals"     YES)
```
Comment (lalitb) [inline CMakeLists.txt:256]: Will leave it for you to decide if the `substrate` keyword is not supposed to be Microsoft internal naming, as this will will be exposed in the public repo? Else, you may want to change the files names to something more generic :)

Comment (leunardo) [reply CMakeLists.txt:256]: @microsoft/1ds-c-substratesignals what do you folks think?

File: Solutions/MSTelemetrySDK.sln (hunk 5 above)
Comment (leunardo) [reply Solutions/MSTelemetrySDK.sln:1291]: Sure, will revert those changes. Visual Studio automatically changed it

**Review by lalitb — APPROVED** [review]
LGTM. Thanks
